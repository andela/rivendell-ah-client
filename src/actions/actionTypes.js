export default {
  SAGA_WORLD: 'SAGA_WORLD',
  LOGIN: 'LOGIN',
  PERSIST_LOGIN: 'PERSIST_LOGIN',
  CLEAR_API_VALIDATION_ERR: 'CLEAR_API_VALIDATION_ERR',
  CLEAR_ALL_API_VALIDATION_ERRS: 'CLEAR_ALL_API_VALIDATION_ERRS',
  LOGGED_IN_USER_PROFILE: 'LOGGED_IN_USER_PROFILE',
  LOGOUT: 'LOGOUT',
  SOCIAL_LOGIN: 'SOCIAL_LOGIN',
  SOCIAL_LOGIN_REDIRECT: 'SOCIAL_LOGIN_REDIRECT',
  CREATE_ARTICLE: 'CREATE_ARTICLE',
  LOAD_CATEGORIES: 'LOAD_CATEGORIES',
  LOAD_ARTICLE: 'LOAD_ARTICLE',
  UPLOAD_IMAGE: 'UPLOAD_IMAGE',
  UPDATE_ARTICLE: 'UPDATE_ARTICLE',
  EDIT_ARTICLE: 'EDIT_ARTICLE',
};
