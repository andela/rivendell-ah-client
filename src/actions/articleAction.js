import types from './actionTypes';
import axiosInstance from '../services/requestHandler';

const getArticle = slug => ({
  type: types.GET_ARTICLE,
  payload: axiosInstance().get(`/articles/${slug}`)
});

export const reportArticleAction = (slug, type, description) => ({
  type: types.REPORT_ARTICLE,
  payload: axiosInstance()
    .post(
      `/articles/${slug}/report`,
      {
        report: { type, description }
      }
    )

});

export default {
  getArticle,
  reportArticleAction,
};
