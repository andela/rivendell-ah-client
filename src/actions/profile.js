import axiosInstance from '../services/requestHandler';
import types from './actionTypes';

const getProfile = path => ({
  type: types.PROFILE,
  payload: axiosInstance().get(path),
});

const updateProfile = (path, data) => ({
  type: types.PROFILE_UPDATE,
  payload: axiosInstance().put(path, {
    user: data,
  }),
});

const updateRedirect = () => ({
  type: types.UPDATE_REDIRECT,
  payload: true,
});

export {
  getProfile,
  updateProfile,
  updateRedirect,
};
