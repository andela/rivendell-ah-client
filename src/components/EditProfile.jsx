import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Form, Button, Container, Image } from 'semantic-ui-react';

/**
 * this function contain the edit profile form
 * @param {Object } formData contain profile detail
 * @param { function } handleSubmitForm
 * @param { function } handleChange
 * @param { function } mouseOutHandler
 * @param { bool } isLoading
 * @returns { object } return the EditProfle component
 */
const EditProfile = ({
  formData,
  handleSubmitForm,
  handleChange,
  mouseOutHandler,
  imageLoading,
  handleImageUpload,
  isLoading }) => {
  const {
    firstName,
    lastName,
    email,
    bio,
    username,
    image
  } = formData;
  let imageEle =
    <Image id="profile-image" src={image} alt={username} circular />;
  if (!image) {
    imageEle = (
      <i className="material-icons" id="alternate-profile-image">
        account_circle
      </i>
    );
  }

  /* eslint-disable jsx-a11y/label-has-for */
  return (
    <Container>
      <div className="content">
        <Form loading={isLoading || imageLoading}>
          <div id="edit-profile-header">
            <h2>Edit your Profile</h2>
          </div>
          <div id="profile-image-div">
            {imageEle}
          </div>
          <div className="form-group">
            <label htmlFor="firstname">
              Firstname
            </label>
            <br />
            <input
              type="text"
              id="firstname"
              className="form-field field-length-ok"
              onChange={handleChange}
              onMouseOut={mouseOutHandler}
              onBlur={mouseOutHandler}
              name="firstName"
              value={firstName}
            />
            <span className="status-box" />
          </div>
          <div className="form-group">
            <label htmlFor="lastname">
              Lastname
            </label>
            <br />
            <input
              type="text"
              id="lastname"
              className="form-field field-length-ok"
              onChange={handleChange}
              onMouseOut={mouseOutHandler}
              onBlur={mouseOutHandler}
              name="lastName"
              value={lastName}
            />
            <span className="status-box" />
          </div>
          <div className="form-group">
            <label htmlFor="email">
              Email
            </label>
            <br />
            <input
              type="email"
              id="email"
              className="form-field field-length-ok valid-email"
              onChange={handleChange}
              onMouseOut={mouseOutHandler}
              onBlur={mouseOutHandler}
              name="email"
              disabled="disabled"
              value={email}
            />
            <span className="status-box" />
          </div>
          <div className="form-group">
            <label htmlFor="photo">
              Photo
            </label>
            <br />
            <input
              id="photo"
              type="file"
              className="form-field"
              onChange={handleImageUpload}
              name="image"
            />
          </div>
          <div className="form-group">
            <label htmlFor="bio">
              Biography
            </label>
            <br />
            <textarea
              id="bio"
              onChange={handleChange}
              onMouseOut={mouseOutHandler}
              onBlur={mouseOutHandler}
              className="form-field field-length-ok"
              name="bio"
              defaultValue={bio}
            />
            <span className="status-box" />
          </div>
          <div>
            <Button
              id="edit-profile-button"
              size="medium"
              fluid
              onClick={handleSubmitForm}
            >
              Update
            </Button>
          </div>
        </Form>
        <div id="edit-profile-footer">
          <Link id="profile-redirect-link" to={`/@${username}`}>Back</Link>
        </div>
      </div>
    </Container>
  );
};

EditProfile.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  handleSubmitForm: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  mouseOutHandler: PropTypes.func.isRequired,
  handleImageUpload: PropTypes.func.isRequired,
  imageLoading: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
};

export default EditProfile;
