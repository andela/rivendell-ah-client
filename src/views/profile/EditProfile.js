import React from 'react';
const EditProfile = () => (
    <div>
        <h2>Edit a Profile</h2>
    </div>
);
export default EditProfile;