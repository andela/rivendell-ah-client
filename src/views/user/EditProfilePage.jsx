import React, { Component } from 'react';
import validator from 'validator';
import { Link, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { updateProfile, updateRedirect } from '../../actions/profile';
import EditProfile from '../../components/EditProfile.jsx';
import mouseOutHandler from '../../helpers/profileHelper';

/**
 * Edit profile page
 * @returns {string} rendered html page
 */
export class EditProfilePage extends Component {
  /**
   * constructor function
   * @returns {object} the state propertiest
   */
  constructor() {
    super();
    this.handleUpdateProfile = this.handleUpdateProfile.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.mouseOutHandler = mouseOutHandler.bind(this);
    this.state = {};
  }

  /**
   * update the state with user profile
   * @returns {object} set formData to state
   */
  componentWillMount() {
    this.setState({
      formData: this.props.userProfile,
    });
  }

  /**
   * @param { object } event element recieving an
   * event
   * @returns { object } the event object
   */
  handleUpdateProfile(event) {
    event.preventDefault();
    this.props.updateProfile('/user', this.state.formData);
    return event;
  }

  /**
   * update state on changes event
   * @param { object } event
   * @returns { object } event
   */
  handleChange(event) {
    event.persist();
    const { name, value, id } = event.target;
    if ((name === 'firstName' || name === 'lastName'
      || name === 'email' || name === 'bio')
      && (value.length === 0 || value.length >= 2)) {
      let eleClass = event.target.className;
      if (eleClass.indexOf('field-length-fail') >= 0) {
        eleClass = eleClass
          .replace('field-length-fail', 'field-length-ok');
        const ele = document.getElementById(id);
        ele.setAttribute('class', eleClass);
        const spanEle = event.target.nextSibling;
        spanEle.innerHTML = '';
      }
    }
    if (name === 'email' && (validator.isEmail(value))) {
      let eleClass = event.target.className;
      if (eleClass.indexOf('invalid-email') >= 0) {
        eleClass = eleClass.replace('invalid-email', 'valid-email');
        const ele = document.getElementById(id);
        ele.setAttribute('class', eleClass);
        const spanEle = event.target.nextSibling;
        spanEle.innerHTML = '';
      }
    }
    this.setState(state => (
      {
        formData: {
          ...state.formData, [name]: value,
        }
      }
    ));
    return event;
  }

  /**
   * render the EditprofilePage component
   * and pass down properties to the editProfile
   * component
   * @returns { string } the rendered profile EditProfile component
   */
  render() {
    /* eslint-disable jsx-a11y/anchor-is-valid */
    const { isLoading, redirect } = this.props;
    if (redirect) {
      this.props.updateRedirect();
      return <Redirect to={`/@${this.props.userProfile.username}`} />;
    }
    return (
      <div id="edit-profile-page" className="content">
        <EditProfile
          handleSubmitForm={this.handleUpdateProfile}
          handleChange={this.handleChange}
          mouseOutHandler={this.mouseOutHandler}
          formData={this.state.formData}
          isLoading={isLoading}
        />
        <Link id="profile-redirect-link" to="/">Back</Link>
      </div>
    );
  }
}

/* eslint-disable react/forbid-prop-types */
EditProfilePage.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  redirect: PropTypes.bool.isRequired,
  updateProfile: PropTypes.func.isRequired,
  updateRedirect: PropTypes.func.isRequired,
  userProfile: PropTypes.object.isRequired,
};

// export default EditProfile;
export const mapStateToProps = (state) => {
  const { userProfile,
    profile,
    isLoading,
    redirect,
  } = state.profile;
  return {
    userProfile,
    profile,
    isLoading,
    redirect,
  };
};

export default connect(
  mapStateToProps,
  { updateProfile, updateRedirect }
)(EditProfilePage);
