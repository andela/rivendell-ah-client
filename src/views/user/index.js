import EditProfile from './EditProfile';
import Profile from './Profile';
import Followers from './Followers'
import Following from './Following'

export default {
    EditProfile,
    Profile,
    Followers,
    Following,
}