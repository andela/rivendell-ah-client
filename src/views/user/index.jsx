import EditProfilePage from './EditProfilePage.jsx';
import ProfilePage from './ProfilePage.jsx';
import Followers from './Followers';
import Following from './Following';
import ForgotPassword from './ResetPassword/ForgotPassword';
import ResetPassword from './ResetPassword/ResetPassword';

export default {
  EditProfilePage,
  ProfilePage,
  Followers,
  Following,
  ForgotPassword,
  ResetPassword,
};
