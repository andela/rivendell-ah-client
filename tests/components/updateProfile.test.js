import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { EditProfilePage, mapStateToProps }from '../../src/views/user/EditProfilePage';
import EditProfile from '../../src/components/EditProfile';
import { updateProfile } from '../../src/actions/profile';
import configureStore from 'redux-mock-store';

const initialState = {
  isLoading: false,
  errors: {
    status: false,
    message: '',
    response: {},
  },
  profile: {},
};


const profile = {
  bio: null,
  email: "reader11@gmail.com",
  firstName: "myfirstname",
  id: 18,
  image: null,
  lastName: "mylastname",
  username: "reader11",
  verified: true,
};

const appState = {
  ...initialState,
  profile: {
    userProfile: profile,
  },
}
document.getElementById = (id) => ({
  setAttribute: jest.fn(),
})

const mockStore = configureStore();

describe('Edit Profile', () => {
  const handleChange = jest.fn();
  const handleUpdateProfile = jest.fn();
  const mouseOutHandler = jest.fn();
  describe('component', () => {
    it('should render without error', () => {
      const wrapper = shallow(<EditProfile
        formData={profile}
        handleChange={handleChange}
        handleSubmitForm={handleUpdateProfile}
        mouseOutHandler={mouseOutHandler}
        />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('page', () => {
    it('should render without error', () => {
      const store = mockStore(initialState)
      const wrapper = shallow(
      <EditProfilePage
        updateProfile={updateProfile}
        store={store}
        userProfile={profile}
        handleChange={handleChange}
        handleSubmitForm={handleUpdateProfile}
      />);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render without error', () => {
      const match = {
        params: {
          username: 'reader11'
        }
      }
      const handleChange = jest.fn();
      const handleUpdateProfile = jest.fn();
      const store = mockStore(initialState)
      const wrapper = mount(
        <MemoryRouter>
          <EditProfile
            updateProfile={updateProfile}
            store={store}
            userProfile={profile}
            formData={profile}
            handleChange={handleChange}
            handleSubmitForm={handleUpdateProfile}
            isLoading={false}
          />
        </MemoryRouter>
      );
      const event = {
        preventDefault: ()=>{},

      }
      jest.spyOn(event, 'preventDefault')
     
      const button = wrapper.find('button');
      expect(button.length).toBe(1);
      button.simulate('click');
      expect(handleUpdateProfile).toHaveBeenCalled();
    });

    it('should render without error', () => {
      const store = mockStore(appState);
      const wrapper = mount(
        <MemoryRouter>
          <EditProfilePage
            updateProfile={updateProfile}
            store={store}
            userProfile={profile}
            handleChange={handleChange}
            handleSubmitForm={handleUpdateProfile}
          />
        </MemoryRouter>
      );
      const divEle = wrapper.find('div');
      const updateButton = wrapper.find('button');
      expect(divEle.find('form').length).toBe(1)
      expect(updateButton.length).toBe(1);
    });
  });

  describe('dom events', () => {
    const store = mockStore(appState);
    const userProfile = {
      bio: 'my bio',
      email: "somebody@gmail.com",
      firstName: "somebody",
      id: 18,
      image: null,
      lastName: "somebody",
      username: "somebody",
      verified: true,
    }

    const history = {
      push: jest.fn(),
    }

    const wrapper = mount(
      <MemoryRouter>
        <EditProfilePage
          updateProfile={updateProfile}
          store={store}
          userProfile={userProfile}
          userRedirect={true}
          history={history}
        />
      </MemoryRouter>
    );
    const editProfilePageWrapper = wrapper.find('EditProfilePage');

    describe('handle update profile', () => {
      const event = {
        preventDefault: ()=>{},
      }
      jest.spyOn(event, 'preventDefault')
      it('should execute handleChange on change', () => {
        const handleUpdate = editProfilePageWrapper.instance().handleUpdateProfile(event);
        expect(handleUpdate.preventDefault).toHaveBeenCalled();
      })
    });

    describe('component will receiveProps', () => {
      const componentWillReceiveProps = jest.fn();
      const nextProps = {
          imageUrl:'true',
          isLoading:true,
          success:true,
      }
      editProfilePageWrapper.instance().componentWillReceiveProps = componentWillReceiveProps;
      editProfilePageWrapper.instance().componentWillReceiveProps(nextProps)
    })
    describe('mouse out handler', () => {
      const event = {
        target: {
          name: 'email',
          value: '',
          className: 'field-length-ok',
          id: 'email',
          nextSibling: {}
        }
      }

      it('should render firstname in input field', () => {
        event.target.value = 'e'
        const handleMouseOut = editProfilePageWrapper.instance().mouseOutHandler(event);
        expect(handleMouseOut.target.name).toBe('email');
        expect(handleMouseOut.target.nextSibling.innerHTML)
        .toEqual('length of email must be greater than 2')
      });

      it('should check changes on email field', () => {
        event.target.value = 'somebody';
        event.target.className = 'valid-email';
        const mouseOutEvent = editProfilePageWrapper.instance().mouseOutHandler(event);
        expect(mouseOutEvent.target.value).toEqual('somebody');
        expect(mouseOutEvent.target.name).toEqual('email');
        expect(mouseOutEvent.target.id).toEqual('email');
        expect(mouseOutEvent.target.nextSibling.innerHTML).toEqual('please enter a valid email');
      })
    });
    describe('handle change', () => {
      const event = {
        target: {
          name: 'email',
          value: '',
          className: 'field-length-ok',
          id: 'email',
          nextSibling: {},
        },
        persist: () => {}
      }
      it('should execute handleChange on change', () => {
        event.target.value = ''
        event.target.nextSibling.innerHTML = '';
        const handleChangeEvent = editProfilePageWrapper.instance().handleChange(event);
        expect(handleChangeEvent.target.value).toEqual('');
        expect(handleChangeEvent.target.className).toEqual('field-length-ok')
        expect(handleChangeEvent.target.id).toEqual('email')
      })

      it('should execute handleChange on change', () => {
        event.target.value = 's';
        event.target.className = 'field-length-fail'
        const handleChangeEvent = editProfilePageWrapper.instance().handleChange(event);
        expect(handleChangeEvent.target.value).toEqual('s');
        expect(handleChangeEvent.target.className).toEqual('field-length-fail')
        expect(handleChangeEvent.target.id).toEqual('email')
      })
      it('should execute handleChange on change', () => {
        event.target.value = '';
        event.target.className = 'field-length-fail'
        const handleChangeEvent = editProfilePageWrapper.instance().handleChange(event);
        expect(handleChangeEvent.target.value).toEqual('');
        expect(handleChangeEvent.target.className).toEqual('field-length-fail')
        expect(handleChangeEvent.target.id).toEqual('email')
      })
      it('should execute if email is valid', () => {
        event.target.value = 'some@body.com';
        event.target.className = 'invalid-email';
        const handleChangeEvent = editProfilePageWrapper.instance().handleChange(event);
        expect(handleChangeEvent.target.value).toEqual('some@body.com');
        expect(handleChangeEvent.target.className).toEqual('invalid-email')
        expect(handleChangeEvent.target.id).toEqual('email')

      })
    });
  })

  describe('mapStateToProps', () => {
    it('should map state to props correctly', () => {
      const appState = {
        profile: {
          isLoading: false,
          errors: {},
          profile: {},
          userProfile: {},

        },
        uploadImageReducer: {
        imageUrl:'true',
        isLoading:true,
        success:true,
      },
      }
      const componentState = mapStateToProps(appState);
      const expectedState = {
        isLoading: false,
        profile: {},
        userProfile: {},
        imageUrl:'true',
        imageLoading:true,
        success:true,
      }
      expect(componentState).toEqual(expectedState);
    });
  });
});
