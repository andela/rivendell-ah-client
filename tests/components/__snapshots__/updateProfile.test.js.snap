// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Profile component should render without error 1`] = `
<Container>
  <div
    className="content"
  >
    <Form
      as="form"
    >
      <div
        id="edit-profile-header"
      >
        <h2>
          Edit your Profile
        </h2>
      </div>
      <div
        id="profile-image-div"
      >
        <i
          className="material-icons"
          id="alternate-profile-image"
        >
          account_circle
        </i>
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="firstname"
        >
          Firstname
        </label>
        <br />
        <input
          className="form-field field-length-ok"
          id="firstname"
          name="firstName"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          onMouseOut={[MockFunction]}
          type="text"
          value="myfirstname"
        />
        <span
          className="status-box"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="lastname"
        >
          Lastname
        </label>
        <br />
        <input
          className="form-field field-length-ok"
          id="lastname"
          name="lastName"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          onMouseOut={[MockFunction]}
          type="text"
          value="mylastname"
        />
        <span
          className="status-box"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="email"
        >
          Email
        </label>
        <br />
        <input
          className="form-field field-length-ok valid-email"
          disabled="disabled"
          id="email"
          name="email"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          onMouseOut={[MockFunction]}
          type="email"
          value="reader11@gmail.com"
        />
        <span
          className="status-box"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="photo"
        >
          Photo
        </label>
        <br />
        <input
          className="form-field"
          id="photo"
          name="image"
          type="file"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="bio"
        >
          Biography
        </label>
        <br />
        <textarea
          className="form-field field-length-ok"
          defaultValue={null}
          id="bio"
          name="bio"
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          onMouseOut={[MockFunction]}
        />
        <span
          className="status-box"
        />
      </div>
      <div>
        <Button
          as="button"
          fluid={true}
          id="edit-profile-button"
          onClick={[MockFunction]}
          role="button"
          size="medium"
        >
          Update
        </Button>
      </div>
    </Form>
    <div
      id="edit-profile-footer"
    >
      <Link
        id="profile-redirect-link"
        replace={false}
        to="/@reader11"
      >
        Back
      </Link>
    </div>
  </div>
</Container>
`;

exports[`Edit Profile page should render without error 1`] = `
<div
  className="content"
  id="edit-profile-page"
>
  <EditProfile
    formData={
      Object {
        "bio": null,
        "email": "reader11@gmail.com",
        "firstName": "myfirstname",
        "id": 18,
        "image": null,
        "lastName": "mylastname",
        "username": "reader11",
        "verified": true,
      }
    }
    handleChange={[Function]}
    handleImageUpload={[Function]}
    handleSubmitForm={[Function]}
    mouseOutHandler={[Function]}
  />
</div>
`;
